torch/overrides.py

```diff
@@ -1785,6 +1787,37 @@ def _no_torch_function_mode() -> Iterator[None]:
         _set_torch_function_mode(old)
 
 
+class TorchFunctionModeInfo(_ModeInfo):
+    def __init__(self):
+        super().__init__(mode_name="torch_function", mode_class=TorchFunctionMode,
+                         base_mode_class=BaseTorchFunctionMode,
+                         required_fn="__torch_function__")
+
+    def is_allowed_type(self, mode) -> bool:
+        return mode is None or\
+            isinstance(mode, self.mode_class) or\
+            (isinstance(mode, type) and not issubclass(mode, self.mode_class))
+
+    def allowed_types_for_error_message(self) -> str:
+        return "TorchFunctionMode, Tensor-like class, or None"
+
+    def get_mode(self):
+        return _get_torch_function_mode()
+
+    def help_text(self, mode) -> str:
+        if isinstance(mode, self.mode_class):
+            return f'Use push_{self.mode_type}_mode instead.'
+        else:
+            return (
+                'If you intended to completely override the preexisting mode, '
+                'pass ignore_preexisting=True.  This can result in unexpected '
+                'behavior; please consider rewriting your mode to be a subclass '
+                f'of {self.mode_class.__name__} to make it compositional!'
+            )
```

Comment:
This error text feels like it should be shared between the two modes, for the most type